Monika-After-Story/MonikaModDev
# Generalization of bookmarks/derand

closes #5616 

This PR implements a generalized framework for bookmarks/derandoms, which allows more use cases, for example in this PR, allowing songs to be derandomed.

## Changes:
- Massive generalization of framework
- `getBookmarkedTopics`/`getDerandomedTopics` now accept arguments specifying which persistent variables to check to get bookmarked/derandomed eventlabels
- Moved the core workings of `mas_topic_rerand` to its own sublabel, generalized for use elsewhere
- Made songs no longer automatically derandom
- Bookmark blacklisting/whitelisting is now handled by event rules

## Testing:
- Verify bookmarks/derand still work as they did before, not allowing any labels w/o the `monika_` prefix and with the other criteria they had before (no topics w/o prompt, must be normal+ [bookmark only])
- Verify songs can be both bookmarked and derandomed
- Verify derandoming songs works as well as rerandoming them (likewise for normal topics)
- Verify that topics which have the bookmark whitelist rule can be bookmarked
- Verify that topics with a bookmark blacklist rule cannot be bookmarked
